### PR TITLE
Add Gather tests to RecSys

### PR DESCRIPTION
**Pull request by bertmaher**
Summary: Some RecSys models gather weights for sparse lengths sum.

Differential Revision: D15482130

### CODE REVIEW

File: tests/unittests/RecommendationSystemTest.cpp [hunk 1]
```diff
@@ -384,12 +383,66 @@ class RecommendationSystemTest : public BackendTest {
...
+      size_t sum =
+          sumOfElements(bindings_.get(lengths[i])->getHandle<int32_t>());
+      auto *indices = mod_.createPlaceholder(
+          ElemKind::Int64ITy, {sum}, "indices" + std::to_string(i), false);
+      fillStableRandomIndex(bindings_.allocate(indices)->getHandle<int64_t>(),
+                            2001, 0, tableSizes[i]);
+
+      // Should be able to pass weights - fix later. Currently, just a
+      // randomized constant.
+      Constant *weights_const = createRandomizedConstant(
```
Comment (rdzhabarov) [inline tests/unittests/RecommendationSystemTest.cpp]: s/weights_const/weightsConst (and other names with _ )

File: tests/unittests/RecommendationSystemTest.cpp [hunk 2]
```diff
@@ -483,6 +553,30 @@ class RecommendationSystemTest : public BackendTest {
...
+
+      EXPECT_EQ(resultTensor->size(), miniBatch);
+
+      auto *concatT = bindings_.get(concatPH);
+      auto concatH = concatT->getHandle();
+      // Check that intermediate concat results didn't overflow.
+      std::cout << "Intermediate concats" << std::endl;
+      concatH.dump();
+      for (int i = 0, e = concatH.size(); i < e; ++i) {
+        EXPECT_LE(fabs(concatH.raw(i)), 100);
```
Comment (rdzhabarov) [inline tests/unittests/RecommendationSystemTest.cpp]: how 100 was selected? any reasoning behind that value

**Comment by rdzhabarov** [comment]
this looks like superset of https://github.com/pytorch/glow/pull/2962

File: tests/unittests/RecommendationSystemTest.cpp [hunk 3]
```diff
@@ -733,3 +803,28 @@ TEST_P(RecommendationSystemTest, RecSys_SLS_Only) {
 
   testSLSQuant();
 }
+
+/// Test gathering weights for SLWS.
+TEST_P(RecommendationSystemTest, RecSys_FP32_Gather_Weights) {
+  ENABLED_BACKENDS(Habana);
```
Comment (jfix71) [inline tests/unittests/RecommendationSystemTest.cpp]: Can we enable on the CPU backend as well?

File: tests/unittests/RecommendationSystemTest.cpp [hunk 4]
```diff
@@ -733,3 +803,28 @@ TEST_P(RecommendationSystemTest, RecSys_SLS_Only) {
...
+
+  testRecSys(/* quantizeSLWS */ false,
+             /* quantizeFC */ false,
+             /* convertToFP16 */ false,
+             /* gatherWeights */ true);
+}
+
+/// Test gathering weights for SLWS.
+TEST_P(RecommendationSystemTest, RecSys_FP32_Medium_Gather_Weights) {
+  ENABLED_BACKENDS(Habana);
```
Comment (jfix71) [inline tests/unittests/RecommendationSystemTest.cpp]: Can we enable on the CPU backend as well?

**Comment by bertmaher** [comment]
Ah sorry, meant to say this is stacked on  #2962.  GitHub export of stacked diffs is meh :)

File: tests/unittests/RecommendationSystemTest.cpp (hunk 2 above)
Comment (bertmaher) [reply tests/unittests/RecommendationSystemTest.cpp]: I think it's derived from the max length and values of the SLS.  To be honest it's mostly checking that we don't end up with inf or nan.

**Pull request closed without merging**